### ISSUE TITLE
Riak CS authentication problem with Fog

### ISSUE THREAD

**Issue by nonsense** [basho/riak_cs#1027]
Hello,

I have been trying to configure Riak CS with Ruby/Fog gem for the past few weeks. I have it working with s3cmd, but for some reason I keep on getting 403 Forbidden responses when I try to use it with Fog.

My .s3cfg is the following

```
[default]
access_key = T_QBGDPU1YSRUXBCWNZN
bucket_location = US
cloudfront_host = cloudfront.amazonaws.com
default_mime_type = binary/octet-stream
delete_removed = False
dry_run = False
enable_multipart = True
encoding = UTF-8
encrypt = False
follow_symlinks = False
force = False
get_continue = False
gpg_command = /usr/bin/gpg
gpg_decrypt = %(gpg_command)s -d --verbose --no-use-agent --batch --yes --passphrase-fd %(passphrase_fd)s -o %(output_file)s %(input_file)s
gpg_encrypt = %(gpg_command)s -c --verbose --no-use-agent --batch --yes --passphrase-fd %(passphrase_fd)s -o %(output_file)s %(input_file)s
gpg_passphrase = 
guess_mime_type = True
host_base = media.domain.net
host_bucket = %(bucket)s.media.domain.net
human_readable_sizes = False
invalidate_on_cf = False
list_md5 = False
log_target_prefix = 
mime_type = 
multipart_chunk_size_mb = 15
preserve_attrs = True
progress_meter = True
proxy_host = 
proxy_port = 0
recursive = False
recv_chunk = 4096
reduced_redundancy = False
secret_key = Q8VWeetQHFPVHEhiP371qQ_N7O6-6h3XV0twPw==
send_chunk = 4096
simpledb_host = sdb.amazonaws.com
skip_existing = False
socket_timeout = 300
urlencoding_mode = normal
use_https = False
verbosity = WARNING
```

This works just fine.

However when I configure Fog with the following, I get Forbidden 403

```
  # Paperclip
  Paperclip.options[:command_path] = '/usr/bin/convert'
  Paperclip::Attachment.default_options[:path] = ':class/:attachment/:id_partition/:style/:hash.:extension'
  Paperclip::Attachment.default_options[:url] = ':s3_domain_url'
  Paperclip::Attachment.default_options[:storage] = :fog
  Paperclip::Attachment.default_options[:fog_credentials] = {
    provider:               "AWS",
    aws_access_key_id:      "T_QBGDPU1YSRUXBCWNZN",
    aws_secret_access_key:  "Q8VWeetQHFPVHEhiP371qQ_N7O6-6h3XV0twPw==",
    host:                   "media.domain.net",
    port:                   80,
    scheme:                 "http"
  }
  Paperclip::Attachment.default_options[:fog_directory] = 'bucketname'
```

I expected both requests with Wireshark, and noticed quiet some differences in s3cmd's request and Rails' request:

```
GET /?delimiter=/ HTTP/1.1
Host: bucketname.media.domain.net
Accept-Encoding: identity
content-length: 0
Authorization: AWS T_QBGDPU1YSRUXBCWNZN:+IE7uduys6v92RaJJ36bSJ3og6o=
x-amz-date: Wed, 03 Dec 2014 15:38:33 +0000
```

which is successful, compared to the one sent from Fog/Paperclip/Rails:

```
PUT /attachments/images/001/000/020/original/86e2c3a5b65198246d99b1bbeb3a4af4c14b4a7c.jpg HTTP/1.1
User-Agent: fog-core/1.25.0
Content-Type: image/jpeg
x-amz-acl: public-read
x-amz-content-sha256: STREAMING-AWS4-HMAC-SHA256-PAYLOAD
x-amz-decoded-content-length: 22887
Content-Encoding: aws-chunked
x-amz-date: 20141203T154124Z
Host: bucketname.media.domain.net
Authorization: AWS4-HMAC-SHA256 Credential=T_QBGDPU1YSRUXBCWNZN/20141203/us-east-1/s3/aws4_request, SignedHeaders=content-encoding;content-type;host;x-amz-acl;x-amz-content-sha256;x-amz-date;x-amz-decoded-content-length, Signature=b9e42e84547eac8583ef18eaa815b4eea74e8c2c00c62760ceb8323c8397a042
Transfer-Encoding: chunked
```

I am pretty sure that this configuration worked previously with Riak CS, so I am not sure why exactly I am getting those 403 Forbidden now... anyone else had similar issues?

Thank you in advance,
Anton

**Comment by kuenishi** [comment]
Looks like it's #897 case. It's just not supported for now. Does Fog have any option to switch to old authorization?

**Comment by kellymclaughlin** [comment]
Fog switched to the v4 authentication scheme in `1.25.0`. I'd try it with an older version and it should work.

**Comment by nonsense** [comment]
Thanks a lot, that's it... I don't know how I missed that.

**Comment by kuenishi** [comment]
As https://github.com/basho/riak_cs/pull/1064 has been merged, Fog will work against Riak CS 2.0.0 with `{auth_v4_enabled, true}`.
